File: java/src/main/java/io/hron/ErrorKind.java
package io.hron;

/** The type of error that occurred during parsing, evaluation, or conversion. */
public enum ErrorKind {
  /** Lexer error - invalid tokens in input. */
  LEX("lex"),
  /** Parser error - invalid syntax. */
  PARSE("parse"),
  /** Evaluation error - cannot compute next occurrence. */
  EVAL("eval"),
  /** Cron conversion error - expression not convertible to cron. */
  CRON("cron");

  private final String value;

  ErrorKind(String value) {
    this.value = value;
  }

  /**
   * Returns the lowercase string representation.
   *
   * @return the kind as a lowercase string
   */
  public String value() {
    return value;
  }

  @Override
  public String toString() {
    return value;
  }
}

File: csharp/Hron/ErrorKind.cs
namespace Hron;

/// <summary>
/// The type of error that occurred during parsing, evaluation, or conversion.
/// </summary>
public enum ErrorKind
{
    /// <summary>Lexer error - invalid tokens in input.</summary>
    Lex,
    /// <summary>Parser error - invalid syntax.</summary>
    Parse,
    /// <summary>Evaluation error - cannot compute next occurrence.</summary>
    Eval,
    /// <summary>Cron conversion error - expression not convertible to cron.</summary>
    Cron
}

public static class ErrorKindExtensions
{
    public static string ToValue(this ErrorKind kind) => kind switch
    {
        ErrorKind.Lex => "lex",
        ErrorKind.Parse => "parse",
        ErrorKind.Eval => "eval",
        ErrorKind.Cron => "cron",
        _ => throw new ArgumentOutOfRangeException(nameof(kind))
    };
}
